tests/utils/stl/test/format.py

```diff
@@ -10,312 +10,328 @@
+            return
+
+        shouldFail = TestType.FAIL in test.testType
+        yield TestStep([shared.execFile], shared.execDir, shared.env, shouldFail)
+
+    def execute(self, test, litConfig):
         try:
-            self.setup(test)
-            pass_var, fail_var = test.getPassFailResultCodes()
-            buildSteps, testSteps = self.getSteps(test, lit_config)
+            self._parseScript(test)
+            if test.result is not None:
+                # TRANSITION: Find a way to do this more cleanly.
+                # Lit asserts that the result has not already been set.
+                result = test.result
+                test.result = None
+                return result
+
+            if test.expectedResult and test.expectedResult.isFailure:
+                failVar = lit.Test.XFAIL
+                passVar = lit.Test.XPASS
+            else:
+                failVar = lit.Test.FAIL
+                passVar = lit.Test.PASS
+
+            buildSetupSteps, buildSteps, handleIsenseRspFile, testSetupSteps, testSteps = \
+                self.getSteps(test, litConfig)
 
-            report = ""
+            report = 'Build setup steps:\n'
+            for step in buildSetupSteps:
+                cmd, out, err, rc = self.runStep(step, litConfig)
+
+                if step.shouldFail and rc == 0:
+                    report += 'Build setup step succeeded unexpectedly.\n'
+                elif rc != 0:
+                    report += 'Build setup step failed unexpectedly.\n'
+
+                report += stl.util.makeReport(cmd, out, err, rc)
+                if (step.shouldFail and rc == 0) or (not step.shouldFail and rc != 0):
+                    litConfig.note(report)
+                    return lit.Test.Result(failVar, report)
+
+            report += 'Build steps:\n'
             for step in buildSteps:
-                cmd, out, err, rc = \
-                    self.build_executor.run(step.cmd, step.work_dir,
-                                            step.file_deps, step.env)
+                cmd, out, err, rc = self.runStep(step, litConfig)
 
-                if step.should_fail and rc == 0:
-                    report += "Build step succeeded unexpectedly.\n"
+                if step.shouldFail and rc == 0:
+                    report += 'Build step succeeded unexpectedly.\n'
                 elif rc != 0:
-                    report += "Build step failed unexpectedly.\n"
+                    report += 'Build step failed unexpectedly.\n'
 
                 report += stl.util.makeReport(cmd, out, err, rc)
-                if (step.should_fail and rc == 0) or \
-                        (not step.should_fail and rc != 0):
-                    lit_config.note(report)
-                    return lit.Test.Result(fail_var, report)
+                if (step.shouldFail and rc == 0) or (not step.shouldFail and rc != 0):
+                    litConfig.note(report)
+                    return lit.Test.Result(failVar, report)

```

Comment:
each of these batches of runSteps are basically copy+paste+tweaked.  Can the duplication be extracted?

```
    def runSteps(self, stageName, steps, litConfig):
        report = stageName + ' steps:\n'
        for step in steps:
            cmd, out, err, rc = self.runStep(step, litConfig)

            if step.shouldFail and rc == 0:
                report += stageName + ' step succeeded unexpectedly.\n'
            elif rc != 0:
                report += stageName = ' step failed unexpectedly.\n'

            report += stl.util.makeReport(cmd, out, err, rc)
            if (step.shouldFail and rc == 0) or (not step.shouldFail and rc != 0):
                litConfig.note(report)
                return lit.Test.Result(failVar, report)

    def execute(self, test, litConfig):
        # ...
        self.runSteps('Build setup', buildSetupSteps, litConfig)
        self.runSteps('Build', buildSteps, litConfig)
        if litConfig.edg_drop:
            self.runSteps('Intellisense response', handleIsenseRspFile, litConfig)
        self.runSteps('Test setup', testSetupSteps, litConfig)
        self.runSteps('Test', testSteps, litConfig)
```

Reply:
Sure, I was thinking of doing that eventually.